File: libs/pulumi/caddy.go
package main

import (
	"github.com/pulumi/pulumi-docker/sdk/v5/go/docker"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func createCaddy(ctx *pulumi.Context, network *docker.Network) error {
	image, err := pullImage(ctx, "caddy", "caddy:latest")
	if err != nil {
		return err
	}

	_, err = docker.NewContainer(ctx, "caddy", &docker.ContainerArgs{
		Image:   image.ImageId,
		Name:    pulumi.String("caddy"),
		Restart: pulumi.String("unless-stopped"),
		Uploads: docker.ContainerUploadArray{
			&docker.ContainerUploadArgs{
				File:   pulumi.String("/etc/caddy/Caddyfile"),
				Source: pulumi.String("Caddyfile"),
			},
		},
		Ports: docker.ContainerPortArray{
			&docker.ContainerPortArgs{
				Internal: pulumi.Int(80),
				External: pulumi.Int(80),
			},
		},
		NetworksAdvanced: docker.ContainerNetworksAdvancedArray{
			&docker.ContainerNetworksAdvancedArgs{
				Name: network.Name,
			},
		},
	})
	if err != nil {
		return err
	}

	return nil
}

File: apps/tandemrobotics/caddy.ts
import * as docker from "@pulumi/docker";

const caddyfile = `:80 {
	reverse_proxy homepage:3000
}

lichtblick.localhost:80 {
	reverse_proxy lichtblick:8080
}
`;

export function createCaddy(network: docker.Network) {
    const image = new docker.RemoteImage("caddy", {
        name: "caddy:2.11.4",
        keepLocally: true,
    });

    new docker.Container("caddy", {
        image: image.imageId,
        name: "caddy",
        restart: "unless-stopped",
        uploads: [{ file: "/etc/caddy/Caddyfile", content: caddyfile }],
        ports: [{ internal: 80, external: 80 }],
        networksAdvanced: [{ name: network.name }],
    });
}
